[epaper_spi] Add support for the Inkplate 2

arunderwood opened the pull request:
# What does this implement/fix?

  Adds support for the [Inkplate 2 e-paper display](https://soldered.com/documentation/inkplate/2/overview/) to the `epaper_spi` component.  The display has 3-color support (black, white, red).

The reference I used was the [Inkplate Arduino Library](https://github.com/SolderedElectronics/Inkplate-Arduino-library).

<img width="3268" height="1916" alt="image" src="https://github.com/user-attachments/assets/d082cb37-9bbe-427d-94d8-0fe3b56b18a9" />

  ## Types of changes

  - [ ] Bugfix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] Code quality improvements to existing code or addition of tests
  - [ ] Other

  **Related issue or feature (if applicable):**

  - https://github.com/esphome/feature-requests/issues/2254

  **Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

  - https://github.com/esphome/esphome-docs/pull/5564
  
  ## Test Environment

  - [x] ESP32
  - [x] ESP32 IDF
  - [ ] ESP8266
  - [ ] RP2040
  - [ ] BK72xx
  - [ ] RTL87xx
  - [ ] nRF52840

  ## Example entry for `config.yaml`:

  ```yaml
  spi:
    clk_pin: GPIO18
    mosi_pin: GPIO23

  display:
    - platform: epaper_spi
      model: inkplate2
      cs_pin: GPIO15
      dc_pin: GPIO33
      reset_pin: GPIO21
      busy_pin: GPIO32
      rotation: 90
      lambda: |-
        // Clear to white background
        it.fill(COLOR_OFF);

        // Draw black circle
        it.filled_circle(70, 52, 30, COLOR_ON);

        // Draw red circle
        it.filled_circle(142, 52, 30, Color(255, 0, 0));
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

arunderwood committed:
Add Inkplate2 support

arunderwood committed:
Remove unneeded delays

arunderwood committed:
Add comments to commands

arunderwood committed:
Remove unneeded wait for power off

arunderwood committed:
better comments

arunderwood committed:
Final cleanup

arunderwood committed:
Merge branch 'dev' into epaper-spi-inkplate2

arunderwood commented on tests/components/epaper_spi/test.esp32-idf-inkplate2.yaml:
I'm a little fuzzy on what my tests should look like here.  Is this correct?

clydebarrow replied:
Leave out all the preconfigured options - just have model. Rotation and lambda are ok, but this is just a build test, it doesn't get run for obvious reasons. It just confirms that the model can be configured and compiles without error.

arunderwood replied:
Should be addressed 👍🏻

clydebarrow commented on esphome/components/epaper_spi/epaper_spi_inkplate2.cpp:
```diff
@@ -0,0 +1,181 @@
+  const uint8_t mask = 1 << bit_position;
+
+  bool is_red = (color.red > 200) && (color.green < 100) && (color.blue < 100);
+  bool is_black = !color.is_on();
```
is_on is true for everything except pure black. I think it's better to use a threshold so that anything less than half white maps to black, this works better with images with grayscales or colours.

clydebarrow replied:
Might be a good idea to rebase on https://github.com/esphome/esphome/pull/11887 - this has a function in epaper_spi.cpp to map colours to monochrome.

clydebarrow commented on esphome/components/epaper_spi/models/inkplate2.py:
```diff
@@ -0,0 +1,56 @@
+    "inkplate2",
+    width=104,
+    height=212,
+    data_rate="1MHz",
```
This seems awfully low. Have you tested with higher speeds?

arunderwood replied:
I had copied this over from the Arduino lib where they limit speed to 1mhz https://github.com/SolderedElectronics/Inkplate-Arduino-library/blob/26e72c3fba5a3e7f3c9e920a17ad127dec3017ee/src/boards/Inkplate2.cpp#L28C1-L28C60

However, I removed the 1MHZ cap and it still work just fine with the default 👍🏻

clydebarrow reviewed (changes requested):
Otherwise looks good!

clydebarrow committed:
Merge branch 'dev' into epaper-spi-inkplate2

arunderwood committed:
Remove unesessary data rate override

Zeik0s commented:
I tried building and flashing this PR on my inkplate, used the example yaml config, but didn't get any change on the display. Is there anything else i missed? Used esphome on this commit: 26607713bb1ba0848100384b9ddecd6d4345ba4d 

I also tried building on b274a7f9775552a22400e60f4a1da77f61027cd3, but no success.

arunderwood commented:
I'm having similar issues, it's only working very intermittently. I suspect I've overlooked something in how I used the state machine and there's a kind of race condition happening with how the display is turned on and off and commands are sent. 

Still working to track that down.

arunderwood committed:
[epaper_spi] Fix Inkplate 2 display update timing

Corrects command sequence to match Soldered Electronics reference implementation. Moves power-on command (0x04) to initialization sequence and repositions stop data transfer command (0x11) immediately before refresh to eliminate 38+ second busy wait delay.

clydebarrow commented:
> I'm having similar issues, it's only working very intermittently.

Have you resolved those issues?

clydebarrow committed:
Merge branch 'dev' into epaper-spi-inkplate2

Zeik0s commented:
So far works, but not sure if it's possible to tune the speed of the e-ink display (i know it is slow at refreshing). 

As a video speaks more for itself: 
https://github.com/user-attachments/assets/0c933a56-60ca-4704-9b55-0562a38b8db2

Lambda:
`   lambda: | -
    // Clear to white background
    it.fill(COLOR_OFF);

      // Draw black circle
      it.filled_circle(70, 52, 30, COLOR_ON);

      // Draw red circle
      it.filled_circle(142, 52, 30, Color(255, 0, 0));`

Zeik0s commented:
Alright, then good to go from my testing perspective 😄

arunderwood commented:
Sweet!  I'm still integrating @clydebarrows [more nuanced approach ](https://github.com/esphome/esphome/blob/dev/esphome/components/epaper_spi/epaper_spi_spectra_e6.cpp#L24-L75)to rendering colors, should have one more upcoming commit for this PR.

clydebarrow committed:
Merge branch 'dev' into epaper-spi-inkplate2
